src/rgw/rgw_cr_rados.h

```diff
@@ -222,6 +238,42 @@ class RGWSimpleRadosReadCR : public RGWSimpleCoroutine {
   }
 };
 
+class RGWSimpleOmapKeysReadCR : public RGWSimpleCoroutine {
```

Comment:
@cbodley ```RGWRadosGetOmapKeysCR``` will return ```-EIO``` rather than ```-ENOENT``` , when we use it in the ```RGWReadDataSyncErrorShardsCR```, this CR will failed since it's super-class ```RGWShardCollectCR``` always expects child CR succeed or return with ```-ENOENT```, so maybe we should use ```RGWSimpleOmapKeysReadCR ``` here, as it's retcode will be ```-ENOENT``` or just do some change on the existing ```RGWRadosGetOmapKeysCR ```?  related code is as below:
````
int RGWShardCollectCR::operate() {
  reenter(this) {
    while (spawn_next()) {
      current_running++;

      while (current_running >= max_concurrent) {
        int child_ret;
        yield wait_for_child();
        if (collect_next(&child_ret)) {
          current_running--;
          if (child_ret < 0 && child_ret != -ENOENT) {
            ldout(cct, 10) << __func__ << ": failed to fetch log status, ret=" << child_ret << dendl;
            status = child_ret;
          }
        }
      }
    }
    while (current_running > 0) {
      int child_ret;
      yield wait_for_child();
      if (collect_next(&child_ret)) {
        current_running--;
        if (child_ret < 0 && child_ret != -ENOENT) {
          ldout(cct, 10) << __func__ << ": failed to fetch log status, ret=" << child_ret << dendl;
          status = child_ret;
        }
      }
    }
    if (status < 0) {
      return set_cr_error(status);
    }
    return set_cr_done();
  }
  return 0;
}
````

Reply:
can you tell where `EIO` is coming from in that case? i don't see much difference between `RGWRadosGetOmapKeysCR` and the new `RGWSimpleOmapKeysReadCR`, except that yours is calling the synchronous version of `ioctx.omap_get_vals()` where `RGWRadosGetOmapKeysCR::send_request()` is using `omap_get_vals2()` with `aio_operate()`

Reply:
@cbodley both ```RGWSimpleOmapKeysReadCR``` and ```RGWRadosGetOmapKeysCR ``` will call ```librados::ObjectReadOperation::omap_get_vals2()```, but ```RGWRadosGetOmapKeysCR``` will call it with rval specified, if the related rados object do not exist, ``` RGWRadosGetOmapKeysCR``` will return with rval, the related code is as below:
```c++
//rgw_cr_rados.cc
int RGWRadosGetOmapKeysCR::send_request() {
  int r = store->get_raw_obj_ref(obj, &ref);
  if (r < 0) {
    lderr(store->ctx()) << "ERROR: failed to get ref for (" << obj << ") ret=" << r << dendl;
    return r;
  }

  set_status() << "send request";

  librados::ObjectReadOperation op;
  op.omap_get_vals2(marker, max_entries, entries, nullptr, &rval);

  cn = stack->create_completion_notifier();
  return ref.ioctx.aio_operate(ref.oid, cn->completion(), &op, NULL);
}
//rgw_cr_rados.h
int request_complete() override {
    return rval;
  }
```
but deep into the call chain of ```RGWSimpleOmapKeysReadCR```, we can find it will call ```omap_get_vals2()``` with rval not specified, if the related rados object do not exist, the return value of ```opeator(oid, &op, &bl)``` will be passed, the related code is as below:
```c++
//librados.cc
int librados::IoCtx::omap_get_vals(const std::string& oid,
                                   const std::string& orig_start_after,
                                   const std::string& filter_prefix,
                                   uint64_t max_return,
                                   std::map<std::string, bufferlist> *out_vals)
{
  bool first = true;
  string start_after = orig_start_after;
  bool more = true;
  while (max_return > 0 && more) {
    std::map<std::string,bufferlist> out;
    ObjectReadOperation op;
    op.omap_get_vals2(start_after, filter_prefix, max_return, &out, &more,
		      nullptr);
    bufferlist bl;
    int ret = operate(oid, &op, &bl);
    if (ret < 0) {
      return ret;
    }
    if (more) {
      if (out.empty()) {
	return -EINVAL;  // wth
      }
      start_after = out.rbegin()->first;
    }
    if (out.size() <= max_return) {
      max_return -= out.size();
    } else {
      max_return = 0;
    }
    if (first) {
      out_vals->swap(out);
      first = false;
    } else {
      out_vals->insert(out.begin(), out.end());
      out.clear();
    }
  }
  return 0;
}
```
in order to analysis the rval passed by osd, we specified the rval in ```  RGWSimpleOmapKeysReadCR```'s call chain and display this rval , just as below:
```c++
[root@0deae4d9bb38 librados]# git diff librados.cc 
diff --git a/src/librados/librados.cc b/src/librados/librados.cc
index e9157ab..d1789d7 100644
--- a/src/librados/librados.cc
+++ b/src/librados/librados.cc
@@ -396,10 +396,12 @@ int librados::IoCtx::omap_get_vals(const std::string& oid,
   while (max_return > 0 && more) {
     std::map<std::string,bufferlist> out;
     ObjectReadOperation op;
+    int val;
     op.omap_get_vals2(start_after, filter_prefix, max_return, &out, &more,
-                     nullptr);
+                     &val);
     bufferlist bl;
     int ret = operate(oid, &op, &bl);
+    std::cout << cpp_strerror(ret) << " : " << cpp_strerror(val) << std::endl;
     if (ret < 0) {
       return ret;
     }
```
in this way, the output is as below:
```c++
[root@0deae4d9bb38 build]# rgw sync status --rgw-zone=zone1 --debug-rgw=20 2>kk
          realm d642c3d5-eea7-4a55-8a17-bf6296cf4548 (earth)
      zonegroup 38a8d75d-c834-4987-a78f-e615328fa663 (us)
           zone 71b1be97-96d1-4e0a-81e2-c5e430a186a7 (zone1)
  metadata sync syncing
                full sync: 0/64 shards
                incremental sync: 64/64 shards
                metadata is caught up with master
....
(2) No such file or directory : (5) Input/output error
(0) Success : (0) Success
(2) No such file or directory : (5) Input/output error
....
      data sync source: 9a4a40f0-9b31-4820-a5a8-6be46155f3e6 (zone0)
                        syncing
                        full sync: 0/128 shards
                        incremental sync: 128/128 shards
                        1 shards are recovering
                        recovering shards: [104]
```
so, if the related rados object do not exit, the rval from osd is always ENO(besides this ,we can also find it if we display th retcode of ```RGWRadosGetOmapKeysCR``` in rgw_data_sync.cc when process bucket shards that previously failed), so maybe we should fix the ```RGWRadosGetOmapKeysCR```, so that it will return with ```-ENOENT``` instead of ```-EIO```

Reply:
how come the rval of ```librados::ObjectReadOperation::omap_get_vals2()``` will be ```-EIO```, we are not exactly clear now, @Liuchang0812  and I will take a deep look into it later on, but if we use ```rados listomapkeys```  on an nonexistent object, the retcode will be ```-ENOENT```, just as below:
```c++
[root@0deae4d9bb38 build]# bin/rados -p zone1.rgw.log listomapkeys non-exist-object 
error getting omap key set zone1.rgw.log/non-exist-object: (2) No such file or directory
```
so maybe the reasonable rval in ```omap_get_vals2()``` should be ```-ENOENT```, so that it can stay in consensus with osds.
@cbodley @yehudasa @oritwas hope you can give some kindly suggestions on it

Reply:
it looks like that EIO is coming from `C_ObjectOperation_decodevals::finish()` in Objecter.h. i see the osd replying to the op with ENOENT, but finish() gets called with r=0 and an empty bufferlist. trying to decode that bufferlist results in EIO

Reply:
please see https://github.com/ceph/ceph/pull/19878